src/metrics/metrics.ts

```diff
@@ -0,0 +1,214 @@
+import type {LogSink} from '@rocicorp/logger';
+
+export interface ReporterOptions {
+  datadogApiKey: string;
+  metrics: Metrics;
+  // intervalSecs defaults to 2 minutes.
+  intervalSecs?: number;
+  abortSignal?: AbortSignal;
+  // If a LogSink is not provided, the Reporter is silent.
+  logSink?: LogSink;
+  // Should probably also take tags (eg, env, version). Will
+  // probalby need to take host and service too for the server side.
+}
+
+// Reporter periodically reports metrics to Datadog. It uses an interval
+// instead of a timer because we want to keep to the desired interval as
+// closely as possible. A typical pattern for sampling a client is for the
+// client to report metrics on a given interval and then rollup metrics on
+// the server on the same interval. The extent to which the interval drifts
+// skews the counts seen by the server.
+export class Reporter {
+  private _metrics: Metrics;
+  private _apiKey: string;
+  private _intervalSecs: number;
+  private _intervalID: ReturnType<typeof setInterval> | 0 = 0;
+  private _abortSignal: AbortSignal | undefined;
+  private _logSink: LogSink | undefined;
+
+  constructor(options: ReporterOptions) {
+    this._metrics = options.metrics;
+    this._apiKey = options.datadogApiKey;
+    this._intervalSecs = options.intervalSecs || 2 * 60;
+    this._abortSignal = options.abortSignal;
+    this._logSink = options.logSink;
+
+    if (this._abortSignal !== undefined) {
+      this._abortSignal.addEventListener('abort', () => {
+        this._stopInterval();
+        this._maybeLogDebug('Metrics Reporter aborted');
+      });
+    }
+
+    this._startInterval();
+  }
+
+  private _startInterval() {
+    if (this._intervalID) {
+      return;
+    }
+
+    // eslint-disable-next-line require-await
+    this._intervalID = setInterval(async () => {
```

Comment:
no async here

Reply:
Not sure why lint didn't flag this? Maybe wrong version of @rocicorp/eslint-config